src/frontend/query_commands.ml

```diff
@@ -511,6 +511,164 @@ let dispatch pipeline (type a) : a Query_protocol.t -> a =
+          | true -> 
+            expression := exp :: !expression
+          | false -> 
+            Ast_iterator.default_iterator.expr self exp
+        in
+        let iterator = 
+          {Ast_iterator.default_iterator with expr}
+        in 
+        (match ppxed_parsetree with
+        | `Interface si -> iterator.signature iterator si
+        | `Implementation str -> iterator.structure iterator str)
+      | false -> Ast_iterator.default_iterator.expr self expr
+    in
+    let signature_item (self : Ast_iterator.iterator) 
+    (item_1 : Parsetree.signature_item) =
+      match check_signatures item_1.psig_desc with
+      | true ->
+        let signature_item(self: Ast_iterator.iterator) 
+        (item_2: Parsetree.signature_item) = 
+        (match check_at_pos item_2.psig_loc && item_1 <> item_2 with
+        | true -> 
+          signature := item_2 :: !signature
+        | false -> 
+          Ast_iterator.default_iterator.signature_item self item_2) 
+        in 
+        let iterator = 
+          {Ast_iterator.default_iterator with signature_item}
+        in 
+        (match ppxed_parsetree with
+        | `Interface si -> iterator.signature iterator si
+        | `Implementation str -> iterator.structure iterator str)
+      | false -> Ast_iterator.default_iterator.signature_item self item_1
+    in
+    let structure_item (self : Ast_iterator.iterator) 
+    (item_1 : Parsetree.structure_item) =
+      (match check_structures item_1.pstr_desc with
+      | true ->
+        let structure_item(self: Ast_iterator.iterator) 
+        (item_2: Parsetree.structure_item) = 
+          (match check_at_pos item_2.pstr_loc && item_1 <> item_2 with
+          | true -> 
+            structure := item_2 :: !structure
+          | false -> 
+            Ast_iterator.default_iterator.structure_item self item_2) 
+        in 
+        let iterator = 
+          {Ast_iterator.default_iterator with structure_item}
+        in 
+        (match ppxed_parsetree with
+        | `Interface si -> iterator.signature iterator si
+        | `Implementation str -> iterator.structure iterator str)
+      | false -> Ast_iterator.default_iterator.structure_item self item_1)
+    in
+    let iterator =
+      { Ast_iterator.default_iterator with signature_item; structure_item; expr }
+    in
+    let _ =
+      match parsetree with
+      | `Interface si -> iterator.signature iterator si
+      | `Implementation str -> iterator.structure iterator str
+    in
+    begin
+      match !signature,!structure,!expression with
+      | signature,[],[] when signature <> [] ->
+        let exp =
+          Pprintast.signature Format.str_formatter (List.rev signature);
+          Format.flush_str_formatter ()
+        in
+        ppx_expansion 
+        ~ppx:exp 
+        ~a_start:(Option.get !atr).loc_start
+        ~a_end:(Option.get !atr).loc_end
+      | [],structure,[] when structure <> [] ->
```

Comment:
You could put a `| [], [], [] -> ...` case first so that you don't need the guarding clauses `when ...`